airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py

```diff
@@ -0,0 +1,148 @@
+            yield record
+
+    def read_records(self, sync_mode, stream_slice: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
+        response = self.google_ads_client.send_request(self.get_query(stream_slice))
+        yield from self.parse_response(response)
+
+
+class IncrementalGoogleAdsStream(GoogleAdsStream, ABC):
+    def __init__(self, start_date: str, conversion_window_days: int, **kwargs):
+        self.conversion_window_days = conversion_window_days
+        self._start_date = start_date
+        super().__init__(**kwargs)
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        stream_state = stream_state or {}
+        start_date = stream_state.get(self.cursor_field) or self._start_date
+
+        return chunk_date_range(start_date=start_date, conversion_window=self.conversion_window_days, field=self.cursor_field)
+
+    @staticmethod
+    def get_date_params(stream_slice: Mapping[str, Any], cursor_field: str, end_date: pendulum.datetime = None):
+        end_date = end_date or pendulum.yesterday()
+        start_date = pendulum.parse(stream_slice.get(cursor_field))
+        if start_date > pendulum.now():
+            return start_date.to_date_string(), start_date.add(days=1).to_date_string()
+
+        end_date = min(end_date, pendulum.parse(stream_slice.get(cursor_field)).add(months=1))
+
+        # Fix issue #4806, start date should always be lower than end date.
+        if start_date.add(days=1).date() >= end_date.date():
+            return start_date.add(days=1).to_date_string(), start_date.add(days=2).to_date_string()
+        return start_date.add(days=1).to_date_string(), end_date.to_date_string()
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        current_stream_state = current_stream_state or {}
+
+        # When state is none return date from latest record
+        if current_stream_state.get(self.cursor_field) is None:
+            current_stream_state[self.cursor_field] = latest_record[self.cursor_field]
+
+            return current_stream_state
+
+        date_in_current_stream = pendulum.parse(current_stream_state.get(self.cursor_field))
+        date_in_latest_record = pendulum.parse(latest_record[self.cursor_field])
+
+        current_stream_state[self.cursor_field] = (max(date_in_current_stream, date_in_latest_record)).to_date_string()
+
+        return current_stream_state
+
+    def get_query(self, stream_slice: Mapping[str, Any] = None) -> str:
+        start_date, end_date = self.get_date_params(stream_slice, self.cursor_field)
+        query = GoogleAds.convert_schema_into_query(
+            schema=self.get_json_schema(), report_name=self.name, from_date=start_date, to_date=end_date, cursor_field=self.cursor_field
+        )
+        return query
+
+
+class AdGroupAdReport(IncrementalGoogleAdsStream):
```

Comment:
How is this different than AdGroupAd? Should we be returning just one of them? (is there any information contained in the other that is not contained in this one? )

Reply:
AdGroupAdReport in AdGroupAd have about 50% of the same data in their streams.
AdGroupAd stream is created in such a way that we get absolutely all information about Ads.
AdGroupAdReport stream has only basic information about the essence of the Ads and also pulls up a lot of information on the metrics and segments that relate to the existing Ads.